Soc

JavadEs12 opened the pull request:
Add classes of the new ShortestPath implementation

JEslamibabaheidari committed:
Add HeapNodeTest

JavadEs12 committed:
Add "ArcNew" which is a new arc class

JavadEs12 committed:
Add "NodeNew" which is new Node class implementation

JavadEs12 committed:
Add Network class

JavadEs12 committed:
Add NetworkBuilder Class

JavadEs12 committed:
Add DijkstraAlgorithm class

JavadEs12 committed:
Modifying program class to run the solution according to new classes

JavadEs12 committed:
Removing all implementation classes and tests

JavadEs12 committed:
Revert "Removing all implementation classes and tests"

This reverts commit bfda743607452a933d7a2869aff1363019335e93.

JavadEs12 committed:
Revert "Modifying program class to run the solution according to new classes"

This reverts commit 750cfe1c813b957f7b22f64caec6b1290fcd2443.

JavadEs12 committed:
Removing  Old implementation classes and tests

JavadEs12 committed:
Revert "Add DijkstraAlgorithm class"

This reverts commit f8a14ef7539a8f094a56942ac633600f22bf1f16.

JavadEs12 committed:
Add Dijkstra Algorithm after reverting

3Dmondo commented on ShortestPath/ShortestPath/Arc.cs:
```diff
@@ -6,39 +6,5 @@ public class Arc
         public string Orig { get; set; }
         public string Dest { get; set; }
         public double Cost { get; set; }
```
This class contains both the information about the topology and the "state" of the shortest path computation (the cost).
while the topology doesn't change, the cost is updated during the computation.
I would disentangle the two, for example you can have a dictionary, indexed by the node key holding the costs, this dictionary may be initialized (re created) each time you call the FindShortestPath method as it represent the state of the computation (and in some sense the result of the path search).

JavadEs12 replied:
Well, this is the Arc class and in the Arc class nothing will change even the Arc's Cost, but maybe this comment suits more to the Node class because,  there we have typology and the cost of the Nodes that is changing continuously.

3Dmondo commented on ShortestPath/ShortestPath/Network.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace ShortestPath
+{
+    public class Network
```
Please rename the file to match the class name

JavadEs12 replied:
could you explain a little more?

3Dmondo commented on ShortestPath/ShortestPath/Network.cs:
```diff
@@ -0,0 +1,17 @@
+        public Dictionary<string,Arc> NetworkArcs = new Dictionary<string, Arc>();
+        public Dictionary<string, Node> NetworkNodes = new Dictionary<string, Node>();
+        public Dictionary<string, List<Arc>> NetworkNodesBackArcs = new Dictionary<string, List<Arc>>();  
+        public Dictionary<string, Node> ExtractedNodes = new Dictionary<string, Node>();
```
ExtractedNodes and heap do not belong to the network, they are a "state" for the algorithm. see my other comment in the Arc class.

3Dmondo commented on ShortestPath/ShortestPath/NetworkBuilder.cs:
```diff
@@ -0,0 +1,56 @@
+{
+    public class NetworkBuilder
+    {
+        public void BuildNetwork(Network network, List<Arc> arcs)
```
NettworkBuilder.BuildNetwork should return an object of type Network, not mofifying an input parameter.

JavadEs12 replied:
I think Chosen names for this class and its method were wrong, because Network object was created in Program class and in this class it's going to be assigned not be created .
NetworkBuilder -> NetworkAssigner (class)
NetworkBuild -> NetworkAssign (metthod)

JavadEs12 committed:
Add ArrayHeap class to create Heap By  arrayList data struture

JavadEs12 committed:
Add ArrayHeap AddNewNode method

JavadEs12 committed:
Add ArrayHeap Test methods

JavadEs12 committed:
Add RemoveRoot class to ArrayHeap class

JavadEs12 committed:
Add RemoveRootTest to ArrayHeapTests and fix RemoveRoot bug

JavadEs12 committed:
Add InterfaceExample

JavadEs12 committed:
Rename AddNewNode method in ArrayHeap to Add

JavadEs12 committed:
Rename RemoveRoot method in ArrayHeap to Remove

JavadEs12 committed:
ExtractedNodes and heap didn't belong to Network, so moved to Dijkstra algoritm to be part of algorithm status

JavadEs12 committed:
Rename NetworkAssigner to NetworkBuilder, Add BuildNetwork method which creates a Network object ,assign  it and retrun it

JavadEs12 committed:
Rename NetworkBuilder object":
networkAssigner -> networkBuilder

also rename :
DijkstraAlgorithm object:
search -> Dijkstra

JavadEs12 committed:
Add Iheap Interface

JavadEs12 committed:
Add HeapFactory class and use it in Dijkstra class to implement heap as interface

3Dmondo commented on ShortestPath/ShortestPath/ArrayHeap.cs:
```diff
@@ -0,0 +1,115 @@
+
+        private void PerculateUp()
+        {
+            if (Heap.Count == 0) { return; }
```
use your count property here, not the Heap.Count

3Dmondo commented on ShortestPath/ShortestPath/ArrayHeap.cs:
```diff
@@ -0,0 +1,115 @@
+                while (true)
+                {
+                    Node Child = Heap[pointer];
+                    Node Parent = Heap[Math.Abs((pointer) / 2)]; //Maybe need to round down 
```
why you need to use the "Math.Abs" here? shouldn't the pointer / 2 always return a number greater or equal to zero?

3Dmondo replied:
I'm not sure if this is the right thing to do; if you run the following code:
```
Console.WriteLine(1 / 2); 
Console.WriteLine(2 / 2);
Console.WriteLine(3 / 2);
Console.WriteLine(4 / 2);
Console.WriteLine(5 / 2);
Console.WriteLine(6 / 2); 
```
You will get:
```
0
1
1
2
2
3
```
while I would have expected:
```
0
0
1
1
2
2
```
meaning that nodes in 1 and 2 have as parent the node in 0
nodes in 3 and 4 have as parent the node in 1
and nodes in 5 and 6 have as parent the node in 2

JavadEs12 replied:
in line 43 pointer initialized by (count-1) so => pointer = count -1 
according to : | (n-1)/2 | and we have  | pointer /2 | 
 and the result are as your expectations.

3Dmondo commented on ShortestPath/ShortestPath/ArrayHeap.cs:
```diff
@@ -0,0 +1,115 @@
+                    {
+                        comparer = Child;
+                        Heap[pointer] = Heap[Math.Abs((pointer) / 2)];
+                        Heap[Math.Abs((pointer) / 2)] = comparer;
```
see my other comment

JavadEs12 replied:
because each nodes children are :
2n+1
2n+2
while each nodes parent is : 
| (n-1)/2 |  
why abstraction of  (n-1)/2 ? because if our node number is " 3 ", (3-1)/2 = 1.5 and we dont have node by this index in our list so we have to round it, that is why I used abstraction to get nodes parent index.

3Dmondo commented on ShortestPath/ShortestPath/ArrayHeap.cs:
```diff
@@ -0,0 +1,115 @@
+                        Heap[pointer] = Heap[Math.Abs((pointer) / 2)];
+                        Heap[Math.Abs((pointer) / 2)] = comparer;
+                    }
+                    pointer = Math.Abs((pointer) / 2);
```
see my other comment

JavadEs12 replied:
Done

3Dmondo commented on ShortestPath/ShortestPath/ArrayHeap.cs:
```diff
@@ -0,0 +1,115 @@
+                    Node comparer;
+                    if (Child.Cost < Parent.Cost)
+                    {
+                        comparer = Child;
```
I would extract this block of code in a method taking as argument the two indices and called swap, e.g.:
```
private void Swap(int left, int right)
{
  var tmp = Heap[left];
  Heap[left] = Heap[right];
  Heap[right] = tmp;
}
```

JavadEs12 replied:
I had Swap method by the name  " SubstituteParentByChild ", but forgot to use for this one, used it also for this swapping and rename it to Swap.

3Dmondo commented on ShortestPath/ShortestPath/ArrayHeap.cs:
```diff
@@ -0,0 +1,115 @@
+            int pointer = 0;
+            while (true)
+            {
+                if (Heap.Count == 0) { break; }
```
use your count property here, not the Heap.Count one

3Dmondo commented on ShortestPath/ShortestPath/ArrayHeap.cs:
```diff
@@ -0,0 +1,115 @@
+                }
+            }
+        }
+        private void SubstituteParentByChild(int parentIndex, int childIndex)
```
Ok, I see that you have a Swap method, you can rename this and use it also in the place where I commented

3Dmondo commented on ShortestPath/ShortestPath/ArrayHeap.cs:
```diff
@@ -0,0 +1,115 @@
+                Node LChild = null;
+                Node RChild = null;
+
+                if ((2 * pointer + 1 <= count - 1)) { LChild = Heap[2 * pointer + 1]; }
```
please double check the arithmetic here, look at my other comment about line 47 (I didn't check myself, it can be right)

3Dmondo commented on ShortestPath/ShortestPath/DijkstraAlgorithm.cs:
```diff
@@ -0,0 +1,89 @@
+
+        private void InitializeHeap(Network network)
+        {
+            heap = new Heap(network.NetworkNodes);
```
Are you sure that you need to add all the nodes with infinite cost at the beginning?

3Dmondo commented on ShortestPath/ShortestPath/InterfaceExample.cs:
```diff
@@ -0,0 +1,54 @@
+﻿using System;
```
This file was just an example for the interfaces topic, you can delete it

JavadEs12 replied:
I kept it to check it again. So I do remove it

3Dmondo commented on ShortestPath/ShortestPath/Network.cs:
```diff
@@ -0,0 +1,15 @@
+    public class Network
+    {
+        public Dictionary<string,Arc> NetworkArcs = new Dictionary<string, Arc>();
+        public Dictionary<string, Node> NetworkNodes = new Dictionary<string, Node>();
```
It seems that this Dictionary is not belonging to the Network class, but it is some state of the algorithm, am I wrong?

JavadEs12 replied:
No, these are Network fields that the other 2 ones were related to the algorithm status that I moved them to DijkstraAlgorithm class (ExtractedNodes Dictionary and heap that later convert it to interface)

JavadEs12 committed:
Rename " SubstituteParentByChild " methods name by " Swap " and use it also for the other operations in ArrayHeap Line 48

JavadEs12 committed:
Remove trial Interface

JavadEs12 committed:
Create "Import" class for importing Raw Data and take advantage of OpenClose principle (SOLID Principles)

JavadEs12 committed:
Add UserInput class and move ReadInput and checkInputAccuracy function from Main method to that class

JavadEs12 committed:
Create a Heap folder and put Heap related files there

JavadEs12 committed:
Create Network folder to put Network related stuff there

JavadEs12 committed:
1- Create Network constructor in Network class
2- and then In Network builder calss instead of creating an object of Network class and send it as argument and assign its properties one by one, added some private fields and save network properties to them and finally create a Network object and assign them to it by Network constructor that just is added

JavadEs12 committed:
Match class and file name
import --> DeserializeRawString

JavadEs12 committed:
Some modification :
convert abstract deserializer to interface
renaming some private fields in NetworkBuilder Class and Network class

JavadEs12 committed:
Removing Approximately all of warnings, except some that are not easy to remove

JavadEs12 committed:
1- Rename:
 DijkstraAlgorithm class --> Shorrtestpath class

2- Create Dijkstra class which only apply shortestPath algorithm and moved update method from old Dijkstra class to this class

JavadEs12 committed:
create import file and put files related to importing and reading user inputs to it

JavadEs12 committed:
Create ShortestPath folder and put shortestPath files there

JavadEs12 committed:
Some cleaning and using some expression instead of statements

JavadEs12 committed:
Add Benchmark Project

JavadEs12 committed:
Improve Benchmark For Heaps

JavadEs12 committed:
Add RemoveBenchmark Class in ShortestPath.Benchmark

JavadEs12 committed:
Remove Nullable sign from wholeproject

JavadEs12 committed:
Benchmark Correction, and fix ArrayHeap Remove method Bug

JavadEs12 committed:
1- Remove " RemoveBenchmark.cs " class in Benchmark project

2- Rename Comparison class to HeapComparison  in Benchmark project

JavadEs12 committed:
Add RandomArrayHeapGenerator in ArrayHeap Tests

JavadEs12 committed:
Put the Tests related to each Heap implementation in its own folder and feed them by a random heap generator which is called " RandomHeapGeneration " in ShortestPath.Tests namespace

JavadEs12 committed:
Add DTOTest class

JavadEs12 committed:
1- Add Validate Method
2- Fix Add method bug

JavadEs12 committed:
Remove a wrong name in
id DEBUG
HeapComparison -> Comparison

JavadEs12 committed:
Improve Heap Tests
1 - Convert RandomDictionaryGenerator to static class and use it directly without creating object
2- Improve ArrayHeap Add Test method

JavadEs12 committed:
Reduce Warnings
